File: Java/TopologicalSort/AlienDictionary.java
import java.util.*;
import java.io.*;

class AlienDictionary {

    public static String alienOrder(String[] words) {
        if(words == null || words.length == 0) {
			return "";
		}

		StringBuilder sortedOrder = new StringBuilder();
		HashMap<Character, Integer> inDegree = new HashMap<>();
		HashMap<Character, ArrayList<Character>> graph = new HashMap<>();

		for(String word : words) {
			for(char ch : word.toCharArray()) {
				inDegree.put(ch, 0);
				graph.put(ch, new ArrayList<>());
			}
		}

		for(int i = 0; i < words.length - 1; i++) {
			String word1 = words[i];
			String word2 = words[i + 1];
            if(word2.length() < word1.length() && word1.startsWith(word2)) {
                return "";
            }
			for(int j = 0; j < Math.min(word1.length(), word2.length()); j++) {
				char parent = word1.charAt(j);
				char child = word2.charAt(j);
				if(parent != child) {
					graph.get(parent).add(child);
                    inDegree.put(child, inDegree.get(child) + 1);
					break;
				}
			}
		}

		// processing all valid starting nodes
		Queue<Character> queue = new LinkedList<>();
		for(Map.Entry<Character, Integer> entry : inDegree.entrySet()) {
			if(entry.getValue() == 0) {
				queue.offer(entry.getKey());
			}
		}

		while(!queue.isEmpty()) {
			char vertex = queue.poll();
			sortedOrder.append(vertex);
			ArrayList<Character> children = graph.get(vertex);
			for(char child : children) {
				inDegree.put(child, inDegree.get(child) - 1);
				if(inDegree.get(child) == 0) {
					queue.offer(child);
				}
			}
		}

		if(sortedOrder.length() != inDegree.size()) {
			return "";
		}

		return sortedOrder.toString();
    }

	public static void main(String[] args) {
		// String words[] = {"ywx", "wz", "xww", "xz", "zyy", "zwz"};
		String words[] = {"zy", "zx"};
		String languageOrder = alienOrder(words);
		System.out.println(languageOrder);
	}
}

File: Python/TopologicalSort/AlienDictionary.py
from collections import deque

def alienOrder(words):
    if len(words) == 0:
        return ""

    in_degree = {}
    graph = {}

    for word in words:
        for char in word:
            in_degree[char] = 0
            graph[char] = []

    for i in range(len(words) - 1):
        w1, w2 = words[i], words[i + 1]
        if len(w2) < len(w1) and w1.startswith(w2):
            return ""
        for j in range(min(len(w1), len(w2))):
            parent, child = w1[j], w2[j]
            if parent != child:
                graph[parent].append(child)
                in_degree[child] += 1
                break

    queue = deque()
    for key in in_degree:
        if in_degree[key] == 0:
            queue.append(key)

    sorted_order = []
    while queue:
        vertex = queue.popleft()
        sorted_order.append(vertex)
        for child in graph[vertex]:
            in_degree[child] -= 1
            if in_degree[child] == 0:
                queue.append(child)

    if len(sorted_order) != len(in_degree):
        return ""

    return ''.join(sorted_order)


if __name__ == '__main__':
	words = ["ywx", "wz", "xww", "xz", "zyy", "zwz"]
	ans = findOrder(words)
	print(ans)